### PR TITLE
Create README.md in demos/spawn

### PR DESCRIPTION
**Pull request by OscarShiang**
Make some description of the spawn demo.

Note:
I have once considered putting `k_get_thread_propery` to explain, but I think it is more about some C++ skill. (like `enable_if`)
So I only add `add_named_thread` and `thread_name` in this file.

### CODE REVIEW

File: demos/spawn/README.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# Spawn
+
+In this demo we are going to show the mechanism that a thread can spawn other threads.
```
Comment (DavidSpickett) [inline demos/spawn/README.md]: "thread can use to spawn"

File: demos/spawn/README.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+To illustrate this demo, we need to introduce the following:
+
+* `add_named_thread` (in `src/user/thread.c`)
+* `thread_name` (in `src/user/thread.c`)
+
+## `add_named_thread`
+
+`add_named_thread` uses the `name` and `worker` we provided to create a thread and initializes all its arguments to `0`.
+
+If the creation successes, the function returns the index of the thread as the return value. If the creation fails, the return value will be set to `INVALID_THREAD`.
```
Comment (DavidSpickett) [inline demos/spawn/README.md]: successes -> succeeds
"the function returns the ID of the thread."
"If creation fails, it returns `INVALID_THREAD`."

File: demos/spawn/README.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+After the creation, the thread starts to wait until it is yielded to.
```
Comment (DavidSpickett) [inline demos/spawn/README.md]: "the new thread waits until"

File: demos/spawn/README.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+## `thread_name`
+
+In AMT, we store the name of a thread in a `Thread` structure. 
+
+When we call this function, it uses the `get_thread_property` syscall to copy the name from the kernel, then stores it into the location pointed to by `name`.
```
Comment (DavidSpickett) [inline demos/spawn/README.md]: For the purposes of the demo I think this is fine.

I did just add all that C++ weirdness but it's not worth going into here, syscall name is enough.

File: demos/spawn/README.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+In AMT, we store the name of a thread in a `Thread` structure. 
+
+When we call this function, it uses the `get_thread_property` syscall to copy the name from the kernel, then stores it into the location pointed to by `name`. 
+
+## Walkthrough
+
+We start from the `setup()` function. The `setup()` function is just a starting point for thread 0 to run, so we directly jump into `spawner()` function.
+
+When we enter the function, it first sets the thread name to `spawner`, then creates 3 threads called `Morgoth`, `Sauron` and `Khamul`. After each thread has been created, we send a message to it. These messages indicate how many turns it should wait to log its greeting.
```
Comment (DavidSpickett) [inline demos/spawn/README.md:28]: If you were wondering about the names, I just googled for obscure Lord of the Rings character names :)

File: demos/spawn/README.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+The table below shows the relation between the messages we sent and the destination threads:
+
+    0  ->  Morgoth
+    1  ->  Sauron
+    2  ->  Khamul
+
+After `spawner` finished thread creating and message sending, it calls `thread_join()` to wait for all threads except itself and yields.
```
Comment (DavidSpickett) [inline demos/spawn/README.md]: finished -> finishes

File: demos/spawn/README.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+Since the message number that each thread got is `0`, `1` and `2`, we will print out the greetings of `Morgoth`, `Sauron` and `Khamul` respectively.
```
Comment (DavidSpickett) [inline demos/spawn/README.md]: "respectively after yielding that many times."

**Review by DavidSpickett — COMMENTED** [review]
> I have once considered putting k_get_thread_propery to explain, but I think it is more about some C++ skill. (like enable_if)

Yeah just naming the syscall is fine.

**Comment by OscarShiang** [comment]
Thanks for the suggestion. I have updated the commit.

**Review by DavidSpickett — APPROVED** [review]
LGTM

**Pull request merged**